### PR TITLE
Viite 779 batch job importing addresses change

### PR DESCRIPTION
**Pull request by hestrompa**
Batch to merge to VIITE-700 after changes on road_address table

### CODE REVIEW

**Comment by pedrosag** [comment]
Tests failing on the following:

[info] ProjectDaoSpec:
[info] - check the removal of project links *** FAILED ***
[info]   java.sql.SQLException: ORA-00060: deadlock detected while waiting for resource
------------------------------------------------------------------------------------------------------------------
[info] ProjectServiceSpec:
[info] - process roadChange data and import the roadLink *** FAILED ***
[info]   2 was not equal to 3 (ProjectServiceSpec.scala:926)

**Comment by hestrompa** [comment]
I think the error in test must be related with the VIITE-700 branch itself, because it contains an outdated version of the original branch. Need to merge VIITE-700 again here in this branch after the changes done in there

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala [hunk 1]
```diff
@@ -301,6 +328,92 @@ class AssetDataImporter {
...
+
+      roadsWithMultipleValues.foreach(road =>{
+        val addresses = RoadAddressDAO.fetchByRoadPart(road.roadNumber, road.roadPartNumber, true)
+        addresses.nonEmpty match {
+          case true => {
+            println(s"updating tie = ${road.roadNumber}, aosa = ${road.roadPartNumber}")
+            addresses.foreach(address =>{
+              roadType(roadsWithMultipleValues.filter { r=> r.roadNumber == road.roadNumber && r.roadPartNumber == road.roadPartNumber }, address) match {
+                case Left(roadType) =>{
+                  sqlu"""UPDATE ROAD_ADDRESS SET ROAD_TYPE = ${roadType.value}, ELY= ${road.elyCode} where ID = ${address.id} """.execute
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: Could be better to do updates in bigger chunks than one row at a time.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala [hunk 2]
```diff
@@ -301,6 +328,92 @@ class AssetDataImporter {
...
+        }
+      })
+    }
+
+  }
+
+  def splitRoadAddresses(roadAddress : RoadAddress, addrMToSplit: Long, roadTypeBefore : RoadType, roadTypeAfter : RoadType, elyCode : Long) : Seq[RoadAddress] = {
+    val splitMValue = roadAddress.startMValue + (roadAddress.endMValue - roadAddress.startMValue) / (roadAddress.endAddrMValue - roadAddress.startAddrMValue) * (addrMToSplit - roadAddress.startAddrMValue)
+    println(s"Splitting road address id = ${roadAddress.id}, tie = ${roadAddress.roadNumber} and aosa = ${roadAddress.roadPartNumber}, on MValue = ${addrMToSplit}")
+    val roadAddressA = roadAddress.copy(
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: Does not take into account the digitization direction: would put AgainstDigitization road addresses into wrong order

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala [hunk 3]
```diff
@@ -301,6 +328,92 @@ class AssetDataImporter {
...
+    }
+
+    withDynTransaction{
+      roadsWithSingleRoadType.foreach(road => {
+        updateRoadWithSingleRoadType(road._1, road._2, road._3, road._4)
+      })
+
+      roadsWithMultipleValues.foreach(road =>{
+        val addresses = RoadAddressDAO.fetchByRoadPart(road.roadNumber, road.roadPartNumber, true)
+        addresses.nonEmpty match {
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: match does nothing: if empty, foreach is not performed.

Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: Additionally does not update the earlier versions of the rows

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala (hunk 2 above)
Comment (hestrompa) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: In here, digitization affects MValues? I thought that is only related with geometry.

Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: It tells about the direction of growth.
Towards Digitization: startM = 0.0 <=> startMAddress, endM <=> endMAddress
Against Digitization: startM = 0.0 <=> endMAddress, endM <=> startMAddress

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala (hunk 1 above)
Comment (tvenhola) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: Not worth it, too low number of rows affected -> performance increase is negligible.

**Comment by tvenhola** [comment]
Let's merge this to 700.

**Pull request merged**